Pull Request #4

AlexanderSparrow opened the pull request.

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 1
а the first presentation

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 1
а the second presentation

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 3
а the third presentation

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 3
а the third presentation (fix some bags)

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 4
а the fourth presentation (fix some bags)

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (json, HTTP and new bags)

Task verification iteration: 0
а the zero presentation (add some bags)

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (json, HTTP, server requests and server responses)

Task verification iteration: 1
а the ашкые presentation (So far, I think everything is perfect)

kesch9 commented on src/service/testUpdateEpicStatusAllNew.java:
```diff
@@ -0,0 +1,4 @@
+package service;
+
+public class testUpdateEpicStatusAllNew {
```
Обязательно
Класс не используется, плюс неверно назван. Удалить класс

AlexanderSparrow replied:
нет такого класса...

kesch9 commented on src/model/Epic.java:
```diff
@@ -1,36 +1,82 @@
+    @Expose
     private final Set<SubTask> subTasks;
 
+    @Expose
```
Обязательно
Никакие аннотации Expose не используем в моделях, сериализуем и десерелизуем все поля

AlexanderSparrow replied:
Expose убраны

kesch9 commented on src/service/BaseHttpHandler.java:
```diff
@@ -0,0 +1,30 @@
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+public class BaseHttpHandler implements HttpHandler {
```
Обязательно
Разнесите новые поля по пакетам

AlexanderSparrow replied:
различные сервисные классы разнесены по соответсвующим пакетам

kesch9 commented on src/service/LocalDateAdapter.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+
+public class LocalDateAdapter extends TypeAdapter<LocalDate> {
```
Обязательно
У вас в моделях нет LocalDate и LocalTime поэтому адаптеры для этих классов не нужны

AlexanderSparrow replied:
убрал лишнее

kesch9 commented on src/Main.java:
```diff
@@ -4,28 +4,45 @@
 
-    public static void main(String[] args) {
-        //InMemoryTaskService inMemoryTaskService = Services.getDefault(); // Создание менеджера задач
+    public static void main(String[] args) throws IOException {
```
Александр, вы опять неверно работаете с гитом, вы не смержили восьмое задание в main, вместо этого в седьмое, поэтому в этом PR правки от восьмого, еще раз надо научиться пользоваться гитом, это очень важная вещь в работе, без гита сейчас не один проект не обходиться

AlexanderSparrow replied:
Спасио, я повторю материал. Но там не учили делать это в Идее. Учту на будущее

kesch9 commented on src/service/EpicHandler.java:
```diff
@@ -0,0 +1,118 @@
+
+public class EpicHandler extends BaseHttpHandler {
+    private final FileBackedTaskService taskService;
+    private final Gson gson;
```
Обязательно
Используйте один из принципов ООП и вынесите общие поля taskService, gson в  BaseHttpHandler. Передавайте эти поля в конструкторе

kesch9 commented on src/service/EpicHandler.java:
```diff
@@ -0,0 +1,118 @@
+                .registerTypeAdapter(LocalDate.class, new LocalDateAdapter())          // Регистрация адаптера для LocalDate
+                .registerTypeAdapter(LocalTime.class, new LocalTimeAdapter())          // Регистрация адаптера для LocalTime
+                .registerTypeAdapter(Duration.class, new DurationAdapter())            // Регистрация адаптера для Duration
+                .registerTypeAdapter(SubTask.class, new SubTaskAdapter())
```
Обязательно
Это обработчик Epic вы добавляете SubTaskAdapter, создавать отдельные адаптеры для моделей не надо, достаточно LocalDateTimeAdapter и DurationAdapter

AlexanderSparrow replied:
Лишние адаптеры убраны

kesch9 commented on src/service/EpicHandler.java:
```diff
@@ -0,0 +1,118 @@
+    private final FileBackedTaskService taskService;
+    private final Gson gson;
+
+    public EpicHandler(FileBackedTaskService taskService) {
```
Обязательно
Передаем TaskService в конструктор, нам без разницы какая реализация будет

AlexanderSparrow replied:
Передали TaskService в конструктор

kesch9 commented on src/service/HttpTaskServer.java:
```diff
@@ -0,0 +1,31 @@
+    private final HttpServer server;
+
+    private final int port=8080;
+    public HttpTaskServer(FileBackedTaskService taskService) throws IOException {
```
Обязательно
Передаем TaskService в конструктор, нам без разницы какая реализация будет. Используем один из принципов ООП полиморфизм

AlexanderSparrow replied:
Передаем TaskService в конструктор

kesch9 replied:
Не подправили в HttpTaskServer

kesch9 commented on src/service/HttpTaskServer.java:
```diff
@@ -0,0 +1,31 @@
+
+    private final HttpServer server;
+
+    private final int port=8080;
```
Лучше сделать константой

AlexanderSparrow replied:
private static final int PORT = 8080;

kesch9 commented on src/service/EpicHandler.java:
```diff
@@ -0,0 +1,118 @@
+            // GET /epics/{id}/subtasks - получение подзадач эпика по id
+            int id = Integer.parseInt(pathParts[2]);
+            Set<SubTask> subtasks = taskService.getAllSubTasksByEpic(id);
+            if (subtasks != null) {
```
Обязательно
У вас не будет null по реализации, либо нужно подправить реализацию, либо убрать эту проверку

AlexanderSparrow replied:
убрал проверку

kesch9 commented on src/service/EpicHandler.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    private int handlePostRequest(HttpExchange exchange) throws IOException {
+        String body = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
```
Обязательно
В Post запросах может быть пустое тело сообщения, эта надо учесть и отправить ошибку в ответ. Если тело пустое, то body будет пустой строкой

AlexanderSparrow replied:
Добавлены проверки на пустое тело и отправка ошибки

kesch9 commented on src/service/EpicHandler.java:
```diff
@@ -0,0 +1,118 @@
+
+    }
+
+    private int extractIdFromPath(String path) {
```
Обязательно
Метод не используется, удалить

AlexanderSparrow replied:
изменил реализацию, исключил дублирование кода

kesch9 commented on src/service/SubtaskHandler.java:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    private int extractIdFromPath(String path) {
```
Обязательно
У вас один и тот же метод в нескольких классах, вынесите его в родительский класс. Используйте наследование и выносите общую логику в родительский класс

AlexanderSparrow replied:
сделано

kesch9 commented on src/service/BaseHttpHandler.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
```
Обязательно
Метод не переопределяем, а делаем класс абстрактным, а наследники его переопределят

AlexanderSparrow replied:
Аннотация убрана

kesch9 replied:
Аннотацию не нужно было убирать, а убрать надо сам метод

kesch9 commented on src/service/SubtaskHandler.java:
```diff
@@ -0,0 +1,105 @@
+                } else {
+                    sendText(exchange, "Подзадача не найдена", 404);
+                }
+                break;
```
Обязательно
Добавить default для отправки ошибки

AlexanderSparrow replied:
выполнен рефакторинг кода. default добавлен

kesch9 commented on src/service/TaskHandler.java:
```diff
@@ -0,0 +1,103 @@
+                } else {
+                    sendText(exchange, "Задача не найдена", 404);
+                }
+                break;
```
Обязательно
Добавить default для отправки ошибки

AlexanderSparrow replied:
выполнен рефакторинг кода. default добавлен

kesch9 reviewed (commented):
Используем ООП для реализации кода

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (json, HTTP, server requests and server responses)

Task verification iteration: 2
а the second presentation (fix some bugs)

kesch9 commented on test/modelTests/EpicFieldCalculationTest.java:
```diff
@@ -0,0 +1,44 @@
+package modelTests;
```
Писать тест не нужно в тестах, если вы пишите тесты для моделей, то создаете такой же пакет, что и в src model и располагаете тесты в этом пакете

kesch9 commented on src/handlers/TaskHandler.java:
```diff
@@ -0,0 +1,103 @@
+
+public class TaskHandler extends BaseHttpHandler {
+
+    public TaskHandler(FileBackedTaskService taskService) {
```
Для TaskHandler не подправили, все еще передаете FileBackedTaskService

kesch9 commented on test/endPointsTests/HttpTaskServiceEpicsEndPointTest.java:
```diff
@@ -0,0 +1,113 @@
+package endPointsTests;
```
Название пакетов обычно не пишут в camelCase, обычно одиннарными словами. Т.е. здесь просто endpoint

kesch9 commented on test/ServiceTests/FileBackedTaskServiceTest.java:
```diff
@@ -0,0 +1,117 @@
+package ServiceTests;
```
Название пакета неверное, пакеты не пишут с большой буквы

kesch9 reviewed (approved):
Approve
